### PR TITLE
draft orgUser query

### PR DESCRIPTION
**Pull request by eliykat**
For discussion only, ignore.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    public OrganizationUserUserDetailsQueryResponse(
+        OrganizationUserUserDetails organizationUserUserDetails,
+        bool twoFactorEnabled
+    )
+    {
+        this.OrganizationUserUserDetails = organizationUserUserDetails;
+        this.TwoFactorEnabled = twoFactorEnabled;
+        // Convert the OrganizationUserUserDetails permissions json string 
+        this.Permissions = CoreHelpers.LoadClassFromJsonData<Permissions>(OrganizationUserUserDetails.Permissions);
```
Comment (eliykat) [inline src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs:38]: The class already has a helper for this: `OrganizationUserUserDetails.GetPermissions()`. Although my preference is for the consuming code to do this as/when required, it's duplicative here.

File: src/Api/AdminConsole/Controllers/OrganizationUsersController.cs [hunk 2]
```diff
@@ -100,7 +103,7 @@ public async Task<OrganizationUserDetailsResponseModel> Get(string id, bool incl
         var response = new OrganizationUserDetailsResponseModel(organizationUser.Item1, organizationUser.Item2);
 
         // Downgrade Custom users with no other permissions than 'Edit/Delete Assigned Collections' to User
-        response.Type = GetFlexibleCollectionsUserType(response.Type, response.Permissions);
+        response.Type.GetFlexibleCollectionsUserType(response.Permissions);
```
Comment (eliykat) [inline src/Api/AdminConsole/Controllers/OrganizationUsersController.cs:106]: You're no longer doing anything with this return value.

File: src/Api/AdminConsole/Controllers/OrganizationUsersController.cs [hunk 3]
```diff
@@ -127,27 +130,27 @@ public async Task<ListResponseModel<OrganizationUserUserDetailsResponseModel>> G
...
+        var responses = new List<OrganizationUserUserDetailsResponseModel>();
+        foreach (var queryResponse in queryResponses)
+        {
+            responses.Add(new OrganizationUserUserDetailsResponseModel(
+                    queryResponse.OrganizationUserUserDetails,
+                    queryResponse.TwoFactorEnabled,
+                    permissions: queryResponse.Permissions
+                )
+            );
+        }
```
Comment (eliykat) [inline src/Api/AdminConsole/Controllers/OrganizationUsersController.cs:153]: automapper is definitely overkill. What about:

```cs
var responses = queryResponses.Select(r => new OrganizationUserUserDetailsResponseModel(queryResponse.OrganizationUserUserDetails, queryResponse.TwoFactorEnabled);

File: src/Api/AdminConsole/Models/Response/Organizations/OrganizationUserResponseModel.cs [hunk 4]
```diff
@@ -85,9 +87,12 @@ public class OrganizationUserDetailsResponseModel : OrganizationUserResponseMode
...
 public class OrganizationUserUserDetailsResponseModel : OrganizationUserResponseModel
 {
+    // Since the OrganizationUserUserDetailsQueryResponse object already has the permissions
+    // converted from json. Add it here as an optional param. If included the base method will
+    // use the permissions if not it will conver
     public OrganizationUserUserDetailsResponseModel(OrganizationUserUserDetails organizationUser,
-        bool twoFactorEnabled, string obj = "organizationUserUserDetails")
-        : base(organizationUser, obj)
+        bool twoFactorEnabled, string obj = "organizationUserUserDetails", Permissions permissions = null)
+        : base(organizationUser, obj, permissions: permissions)
```
Comment (eliykat) [inline src/Api/AdminConsole/Models/Response/Organizations/OrganizationUserResponseModel.cs:95]: Again, we avoid all of this if we omit it from the query and let consumers decide if/when they want to deserialize this.

File: src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        // Should the query authorize? Or should this be the responsibility of the controller? 
+        // Maybe having the query authorize would ensure safety?
+        // Code from the controller for auth:
+        // var authorized = (await _authorizationService.AuthorizeAsync(
+        //     User, OrganizationUserOperations.ReadAll(orgId))).Succeeded;
+        // if (!authorized)
+        // {
+        //     throw new NotFoundException();
+        // }
```
Comment (eliykat) [inline src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs:66]: For now at least, I think this should continue to live in the controller; our private and public api use different authorization patterns (private uses a mix of handrolled and resource-based authorization, public uses policy-based authorization), and introducing them at the query level tightly couples the query logic with this particular authorization pattern.

Maybe there's some opportunity to align this in the future, I'm not sure.

File: src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            .GetManyDetailsByOrganizationAsync(request.OrganizationId, request.IncludeGroups, request.IncludeCollections);
+
+        var responseTasks = organizationUsers
+            .Select(async o =>
+            {
+                var orgUser = new OrganizationUserUserDetailsQueryResponse(o,
+                    await _userService.TwoFactorIsEnabledAsync(o));
+
+                // Using the new extension method
+                orgUser.OrganizationUserUserDetails.Type.GetFlexibleCollectionsUserType(orgUser.Permissions);
```
Comment (eliykat) [inline src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs:78]: Unused return value

**Review by eliykat — COMMENTED** [review]
All looks pretty good to me, a few comments below but nothing major. Please feel free to put up a PR for review whenever you're ready.

File: src/Api/AdminConsole/Controllers/OrganizationUsersController.cs (hunk 2 above)
Comment (ttalty) [reply src/Api/AdminConsole/Controllers/OrganizationUsersController.cs:106]: haha yea I missed that part. Thanks.

File: src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs (hunk 1 above)
Comment (ttalty) [reply src/Api/AdminConsole/Queries/OrganizationUserUserDetailsQuery.cs:38]: I don't know how I missed that. That is 100 times easier. Thanks.

**Pull request closed without merging**